### PR TITLE
fix: search timeout

### LINKED ISSUE #10731: `/api/v2/search` timeouts
## Description

https://eth.blockscout.com/api/v2/search?q=uniswap results in a timeout, while https://eth.blockscout.com/api/v2/search/quick?q=uniswap returns a response within several seconds.

## Version

[v6.8.0-beta.+commit.804b9f66](https://github.com/blockscout/blockscout/commit/804b9f66) (first noticed in 6.7.2)

### PR DESCRIPTION
**Pull request by nikitosing**
Closes #10731 
## Motivation
pagination in search results doesn't work: 
1. go to https://eth.blockscout.com/search-results?q=usdt
2. click to the next page
3. click again, and see, that the same tokens are on the page

And slow queries (#10731)

## Changelog
- Improved query performance with splitting tokens query to 2: certified and not certified
- total refactoring of search module 
## Checklist for your Pull Request (PR)

- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
- [ ] I checked whether I should update the docs and did so by submitting a PR to [docs repository](https://github.com/blockscout/docs).
- [ ] If I added/changed/removed ENV var, I submitted a PR to [docs repository](https://github.com/blockscout/docs) to update the list of [env vars](https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md) and I updated the version to `master` in the Version column. If I removed variable, I added it to [Deprecated ENV Variables](https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables/deprecated-env-variables/README.md) page. After merging docs PR, changes will be reflected in these [pages](https://docs.blockscout.com/for-developers/information-and-settings/env-variables).
- [ ] If I added new DB indices, I checked, that they are not redundant, with PGHero or other tools.
- [ ] If I added/removed chain type, I modified the Github CI matrix and PR labels accordingly.

## Release Notes

- **New Features**
	- Enhanced search functionality with improved handling of queries and results.
	- Streamlined pagination logic for search results.

- **Bug Fixes**
	- Adjusted handling of AJAX requests in block data responses.

- **Documentation**
	- Updated module documentation for clarity and improved structure.

- **Tests**
	- Expanded test coverage for search functionality and pagination scenarios.

- **Refactor**
	- Refined internal logic for search and pagination functions to enhance modularity and readability.

### CODE REVIEW

File: apps/explorer/lib/explorer/chain/search.ex [hunk 1]
```diff
@@ -24,126 +22,200 @@ defmodule Explorer.Chain.Search do
...
+  @contract_sorting [
+    {:desc_nulls_last, :certified, :smart_contract},
+    {:asc, :name, :smart_contract},
+    {:desc, :inserted_at, :smart_contract}
+  ]
+
+  @label_sorting [{:asc, :display_name, :address_tag}, {:desc, :inserted_at, :address_to_tag}]
+
   @doc """
     Search function used in web interface. Returns paginated search results
```
Comment (vbaranov) [inline apps/explorer/lib/explorer/chain/search.ex:52]:
```suggestion
    Search function used in web interface and API v2. Returns paginated search results
```

File: apps/explorer/lib/explorer/chain/search.ex [hunk 2]
```diff
@@ -153,122 +225,117 @@ defmodule Explorer.Chain.Search do
       2. Results couldn't be paginated
   """
```
[the remark this answers is not in the document]
Comment (nikitosing) [reply apps/explorer/lib/explorer/chain/search.ex:228]: Processed in 9ec7eae78c06386dda72b21a1b15074b3f86555c

File: apps/explorer/lib/explorer/chain/search.ex (hunk 1 above)
Comment (nikitosing) [reply apps/explorer/lib/explorer/chain/search.ex]: Processed in 9ec7eae78c06386dda72b21a1b15074b3f86555c

File: apps/explorer/lib/explorer/chain/search.ex [hunk 3]
```diff
@@ -675,4 +824,128 @@ defmodule Explorer.Chain.Search do
...
+    inserted_at_datetime = DateTime.to_iso8601(inserted_at)
+
+    %{
+      "certified" => certified,
+      "name" => name,
+      "inserted_at" => inserted_at_datetime
+    }
+  end
+
+  def parse_paging_options(%{"next_page_params_type" => "search"} = paging_params) do
```
[the remark this answers is not in the document]
Comment (nikitosing) [reply apps/explorer/lib/explorer/chain/search.ex:942]: Processed in 9ec7eae78c06386dda72b21a1b15074b3f86555c

File: apps/explorer/lib/explorer/chain/search.ex [hunk 4]
```diff
@@ -675,4 +824,128 @@ defmodule Explorer.Chain.Search do
...
+          Map.put(acc, type, paging_options(paging_params[type]))
+        else
+          acc
+        end
+      end)
+
+    [paging_options: %{default_paging_options() | key: key}]
+  end
+
+  def parse_paging_options(_) do
```
[the remark this answers is not in the document]
Comment (nikitosing) [reply apps/explorer/lib/explorer/chain/search.ex:955]: Processed in 9ec7eae78c06386dda72b21a1b15074b3f86555c

File: apps/explorer/lib/explorer/chain/search.ex [hunk 5]
```diff
@@ -311,139 +380,144 @@ defmodule Explorer.Chain.Search do
...
-          select: ^token_search_fields
-        )
-    end
+  defp search_token_query_certified(term, paging_options) do
+    term
+    |> base_token_query(paging_options)
+    |> where([smart_contract: smart_contract], smart_contract.certified)
+  end
+
+  defp base_token_query(term, paging_options) do
```
[the remark this answers is not in the document]
Comment (nikitosing) [reply apps/explorer/lib/explorer/chain/search.ex]: Renamed!

File: apps/explorer/lib/explorer/chain/search.ex [hunk 6]
```diff
@@ -604,12 +750,14 @@ defmodule Explorer.Chain.Search do
 
   defp search_ens_name(search_query, options) do
     if result = search_ens_name_in_bens(search_query) do
+      {ens_result, address_hash} = result
+
```
[the remark this answers is not in the document]
Comment (nikitosing) [reply apps/explorer/lib/explorer/chain/search.ex:761]: Nice suggestion, thanks

**Review by sl1depengwyn — APPROVED** [review]

**Review by vbaranov — APPROVED** [review]

**Pull request merged**